Add featured image to Media & Text block

carolinan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Add an option to use the featured image in a Media & Text block.
Closes https://github.com/WordPress/gutenberg/issues/35221

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
The Media & Text block has a unique feature called "stack on mobile" which is useful when you want to display the post title and featured image together but with different layouts on desktop and mobile.
This was a feature requested for the "Blue Note" community theme.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
- Adds a new block attribute called "useFeaturedImage". Adds the post id and post type as a new context.
- Deprecates the version that does not have a featured image option
- Adds the featured image option in edit.js and updates the conditions for showing the block options.
- Adds an index.php for outputting the block on the front.
- Updates the media-container utility file to include the featured image option.

The featured image can be selected in the initial block placeholder and from the block toolbar.
A placeholder image is used if the featured image option is chosen, but the post has no featured image.

## Testing Instructions

1. Open the block editor and add a media and text block

Test the following:
Replacing an existing image or video with a featured image
Adding a featured image when there is no featured image on the post or page
Adding a featured image when there is a featured image on the post or page

Removing the featured image from the post/page sidebar
Replacing the featured image from the post/page sidebar
Replacing the featured image from the block toolbar: by uploading a new media and by selecting from the media library

Next, please confirm that:

- Alt text works
- The cropping option and the focal point picker works
- Styling and aligning the block works
- Linking the image works

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

In this video, I start with a post that has a featured image. I add a new media & text block, enable the featured image option, and try some block toolbar settings. Then I replace the featured image with an image from the media library and then replace that image with the featured image again.

https://github.com/WordPress/gutenberg/assets/7422055/bb0e6d92-1776-4e05-bf3d-5263303e62ac

In this video, I am enabling the featured image option even though the post has no featured image, showing the placeholder:

https://github.com/WordPress/gutenberg/assets/7422055/99970fbb-c4df-4de1-a9b9-e167096bc19c

In this video, I have added the featured image and an anchor to the block, and I am checking that the anchor is added on the front.

https://github.com/WordPress/gutenberg/assets/7422055/f1c15063-7c0d-45a5-bbc2-a313ffadcbe0

Twenty Twenty Three blog / Home template with featured image and post title in the media & text block inside the query:


https://github.com/WordPress/gutenberg/assets/7422055/f27aa804-04a7-4f2a-835b-18f40be31952

carolinan committed:
Add featured image to Media & Text block

carolinan committed:
Try adding the featured image option to the toolbar (replace flow)

carolinan commented:
@draganescu Hi. If possible, I would like your help with correctly adding the "Use featured image" option to the toolbar in the media replacement flow for the Media & Text block.

With the PR applied, if I add the Media & Text block, select an image from the media library, then select "Replace" and "Use Featured Image" from the block toolbar, I get this warning in the console:
`Warning: A component is changing an uncontrolled input to be controlled. This is likely caused by the value changing from undefined to a defined value,`

(Note there is no 6.3 rush for this feature, no hurry).

carolinan committed:
fix JS warning in templates where there is no featured image.

carolinan committed:
Add deprecation, update fixtures

carolinan commented:
As usual, I am not sure if this deprecation and fixture update is correct.

Another thing to consider is whether an extra CSS class should indicate that the featured image is used. In that case, the deprecation and fixtures need to be updated again.

draganescu commented:
👋🏻 @carolinan 👀 looking at this

draganescu commented:
I see you solved the console error so I took this for a spin, it seems there's a bit of reactiveness missing:




https://github.com/WordPress/gutenberg/assets/107534/5e631de9-0721-4fe7-8247-0bb82da7db51

draganescu commented on packages/block-library/src/media-text/deprecated.js:
```diff
@@ -233,11 +245,124 @@ const v6Supports = {
+			</div>
+		);
+	},
+	migrate: migrateDefaultAlign,
```
I think we don't need this if there is no change in alignment?

draganescu commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -148,9 +153,35 @@ function MediaTextEdit( { attributes, isSelected, setAttributes } ) {
+		setAttributes( {
+			imageFill: false,
+			mediaType: 'image',
+			mediaId: featuredImage,
```
this is different compared to the cover block - we save the actual URL of the featured image here, and we should not, the idea is to flag the block as using the featured imahe but the media id and url should be undefined.

carolinan replied:
When I set `mediaUrl` to undefined, selecting the "Use featured image" option does not work.

carolinan replied:
The only way I could think of, to work around not saving the actual URL, was to introduce an index.php file
that fetches and displays the featured image data on the front.

But this lead to other issues, because there are several block attributes that are not available when registering the block using PHP, in addition to the text-based content, the inner blocks. I am not sure if I should try to work with the `innerHTML` / content, try to use WP_HTML_Tag_Processor, or change the attributes.

draganescu replied:
I believe we _have_ to make the block dynamic in order for it to support featured image. I suggest lean in WP_HTML_Tag_Processor and alter what the save function provides while rendering.

carolinan replied:
I keep thinking that I can *get content* with the tag processor but no  :) 
It doesn't quiet do what I want it to do.

carolinan commented:
> I see you solved the console error

I still see the warning if I first select an image from the media library, and try to replace it with a featured image.

draganescu commented:
> I still see the warning if I first select an image from the media library, and try to replace it with a featured image.

I don't - in Firefox, on a post that _has_ a featured image, adding some image from the media library, then clickin replace > use featured image, shows the featured image, no warning in the console.

carolinan committed:
Try adding index.php

carolinan commented:
In [79bbc27](https://github.com/WordPress/gutenberg/pull/51491/commits/79bbc27866739b3f64ef62ce9a1ec46ec13ee628) I have set 			mediaId, mediaUrl, and mediaAlt to undefined when the featured image option is selected.
It is now possible to change the featured image in the block through the Post settings > Featured Image.

I have added an index.php file for outputting the featured image on the front.

**Known issues:**
1. Anchor does not work. -And disabling the anchor for existing blocks can cause problems for existing sites.
2. Custom links, link target, and rel settings do not work
3. The image resolution setting is unavailable in the editor, so the featured image is always "full".

**Untested, but potential issues**
Media & text block with featured image inside a query block or template.

carolinan commented:
How can I make this block into a dynamic block and still keep the existing anchor block support?

draganescu commented:
@carolinan the anchor support is [not available for dynamic blocks](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/#anchor).

Unfortunately in the rendered the value of the anchor support is unavailable - as far as I've seen - therefore we need to change how the rendered builds the markup: instead of building all the markup, leave what comes from the save function and only replace the image with the featured image (see the cover block for example).

carolinan commented:
@draganescu Thank you, I think I understand. I have got it mostly working now; <del>the part that I am missing is the position information from the focal point picker, when the "crop image to fill" setting is enabled</del>.

carolinan committed:
Index.php: use the $content and only replace the image source and the image background position.

carolinan committed:
Update index.php

carolinan committed:
Update index.php

carolinan committed:
Add alt and class names to the image tag

carolinan committed:
Fix white space issue

carolinan committed:
Add a placeholder for the featured image

draganescu commented on packages/block-library/src/media-text/deprecated.js:
```diff
@@ -208,7 +220,7 @@ const v4ToV5Supports = {
 };
 
-const v6Supports = {
+const v6ToV7Supports = {
```
What is this for? Just adding one more attribute should require no migration or deprecation. Or does it?

carolinan replied:
Deprecations are not my strength.

- If I remove the deprecation, place the block using trunk, and then switch to this branch (and build it, of course), I don't see a block validation error.
- The docs say that I need to add a deprecation **when changing the attributes of a static block.**
- Looking at the history, I added the deprecation _before_ adding the index.php file.
- The save file is unchanged.

draganescu commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -252,24 +286,30 @@ function MediaTextEdit( { attributes, isSelected, setAttributes } ) {
 					__nextHasNoMarginBottom
 					label={ __( 'Alternative text' ) }
-					value={ mediaAlt }
+					value={ mediaAlt || featuredImageAlt }
```
If we use  `featuredImageAlt` and ` featuredImageURL` like this we need to condition them on `useFeaturedImage` being `true` [here](https://github.com/WordPress/gutenberg/pull/51491/files#diff-799bc423023028d1e733982e9d106233470b262e34737b34c559ab7b14e001d7R171), because there can be that mediaURL is false but there is a featured image and we render the wrong thing.

The gist is if `useFeaturedImage` is false we don't care about anything featured image :)

draganescu reviewed (commented):
Good work here 🙇🏻  This tests well, found this styling issue for featured image placeholder state.

![Screenshot 2023-09-26 at 11 34 57](https://github.com/WordPress/gutenberg/assets/107534/6e89a8bc-1c73-4649-b042-7d7908ff6d02)

Possibly a more generous min height may work?

Left a couple of comments:

- why do we need a deprecation/migration or anything?
- a code nit, that doesn't seem to produce a bug but is susceptible to do so in the future.

carolinan commented:
Testing this on the 6.4 nightly, I see a new JavaScript warning that I can not reproduce on 6.3x.

1. Deactivate Gutenberg
2. In the editor, create a new post. Make sure the post does not have a featured image.
3. Add a media and text block with an image. 
4. Activate Gutenberg (with this pr applied).
5. Select the media & text block, and in the block toolbar, select the option "use featured image".

In the console, I see the following:
` 'react-dom.js?ver=18:73 Warning: A component is changing an uncontrolled input to be controlled. This is likely caused by the value changing from undefined to a defined value, which should not happen. Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://reactjs.org/link/controlled-components`

<details>

```
jquery-migrate.js?ver=3.4.1:104 JQMIGRATE: Migrate is installed with logging active, version 3.4.1
react-dom.js?ver=18:29850 Download the React DevTools for a better development experience: https://reactjs.org/link/react-devtools
index.js:77 wp.blockEditor.__experimentalImageSizeControl is deprecated since version 6.3. Please use wp.blockEditor.privateApis.DimensionsTool and wp.blockEditor.privateApis.ResolutionTool instead.
deprecated @ index.js:77
ImageSizeControl @ index.js:34
renderWithHooks @ react-dom.js?ver=18:16315
mountIndeterminateComponent @ react-dom.js?ver=18:20084
beginWork @ react-dom.js?ver=18:21597
beginWork$1 @ react-dom.js?ver=18:27436
performUnitOfWork @ react-dom.js?ver=18:26567
workLoopSync @ react-dom.js?ver=18:26476
renderRootSync @ react-dom.js?ver=18:26444
performSyncWorkOnRoot @ react-dom.js?ver=18:26095
flushSyncCallbacks @ react-dom.js?ver=18:12052
(anonymous) @ react-dom.js?ver=18:25661
react-dom.js?ver=18:73 Warning: A component is changing an uncontrolled input to be controlled. This is likely caused by the value changing from undefined to a defined value, which should not happen. Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://reactjs.org/link/controlled-components
    at input
    at span
    at FormToggle (http://64.local/wp-content/plugins/gutenberg/build/components/index.min.js?ver=5c709f4107d418a0596a:49161:5)
    at div
    at http://64.local/wp-content/plugins/gutenberg/build/components/index.min.js?ver=5c709f4107d418a0596a:669:66
    at UnconnectedHStack (http://64.local/wp-content/plugins/gutenberg/build/components/index.min.js?ver=5c709f4107d418a0596a:50941:71)
    at div
    at http://64.local/wp-content/plugins/gutenberg/build/components/index.min.js?ver=5c709f4107d418a0596a:669:66
    at div
    at http://64.local/wp-content/plugins/gutenberg/build/components/index.min.js?ver=5c709f4107d418a0596a:669:66
    at BaseControl (http://64.local/wp-content/plugins/gutenberg/build/components/index.min.js?ver=5c709f4107d418a0596a:34508:3)
    at ToggleControl (http://64.local/wp-content/plugins/gutenberg/build/components/index.min.js?ver=5c709f4107d418a0596a:65939:3)
    at div
    at UnforwardedPanelBody (http://64.local/wp-content/plugins/gutenberg/build/components/index.min.js?ver=5c709f4107d418a0596a:58482:5)
    at RegisterResetAll (http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:40877:3)
    at ToolsPanelInspectorControl (http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:40895:3)
    at StyleProvider (http://64.local/wp-content/plugins/gutenberg/build/components/index.min.js?ver=5c709f4107d418a0596a:64420:5)
    at Fill (http://64.local/wp-content/plugins/gutenberg/build/components/index.min.js?ver=5c709f4107d418a0596a:62975:3)
    at Fill
    at InspectorControlsFill
    at StyleProvider (http://64.local/wp-content/plugins/gutenberg/build/components/index.min.js?ver=5c709f4107d418a0596a:64420:5)
    at InspectorControlsFill (http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:40844:3)
    at MediaTextEdit (http://64.local/wp-content/plugins/gutenberg/build/block-library/index.min.js?ver=c73b3a787c32b5dadb2b:32025:3)
    at Edit (http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:14029:5)
    at http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:56895:11
    at http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:57094:96
    at http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:59004:113
    at http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:62028:30
    at http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:59576:102
    at http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:61586:11
    at http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:60639:11
    at http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:58766:65
    at http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:57721:24
    at http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:57891:5
    at http://64.local/wp-content/plugins/gutenberg/build/edit-post/index.min.js?ver=44c4cd804030b58817da:7874:5
    at http://64.local/wp-content/plugins/gutenberg/build/data/index.min.js?ver=f0a463ca6734f531afd3:2029:82
    at http://64.local/wp-content/plugins/gutenberg/build/data/index.min.js?ver=f0a463ca6734f531afd3:2149:77
    at WithMultipleValidation(WithToolbarControls(WithBlockHooks(WithToolbarControls(WithInspectorControls(WithInspectorControls(WithDuotoneControls(WithToolbarControls(WithInspectorControl(WithInspectorControl(WithToolbarControls(Edit))))))))))) (http://64.local/wp-content/plugins/gutenberg/build/compose/index.min.js?ver=e3ebae5a25ea97355f8c:2500:10)
    at FilteredComponentRenderer (http://64.local/wp-content/plugins/gutenberg/build/components/index.min.js?ver=5c709f4107d418a0596a:51616:9)
    at EditWithGeneratedProps (http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:14047:5)
    at BlockEdit (http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:14122:5)
    at BlockCrashBoundary (http://64.local/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=c3b9afb71ff5fcf75f8b:15037:5)
    at BlockListBlock (http://64.local/wp-content/plugins/gutenberg/b
printWarning @ react-dom.js?ver=18:73
error @ react-dom.js?ver=18:47
updateWrapper @ react-dom.js?ver=18:1800
updateProperties @ react-dom.js?ver=18:10161
commitUpdate @ react-dom.js?ver=18:11057
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24399
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24356
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24356
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24356
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24356
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24356
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24466
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24342
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24342
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24342
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24342
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24342
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24356
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24356
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24466
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24356
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24356
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24356
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24586
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24356
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24303
recursivelyTraverseMutationEffects @ react-dom.js?ver=18:24283
commitMutationEffectsOnFiber @ react-dom.js?ver=18:24356
```


</details>

carolinan commented:
When testing further, I see this warning on the Gutenberg trunk too, for actions like toggling the "Crop image to fill entire column"  option.

carolinan committed:
Remove the deprecation

carolinan commented:
I can reproduce the strange height of the placeholder, I would like to understand what has changed between 6.3 and 6.4 since this is a new bug. 🤔

carolinan committed:
Add condition with useFeaturedImage, remove alt text option from placeholder

carolinan commented:
is it because this has not been handled in the block? 🤔 

`wp.blockEditor.__experimentalImageSizeControl is deprecated since version 6.3. Please use wp.blockEditor.privateApis.DimensionsTool and wp.blockEditor.privateApis.ResolutionTool instead.`

carolinan commented:
The JS warning about the uncontrolled input has been solved, the height issue remains.

carolinan committed:
Set a minimum height for the placeholder image

MaggieCabrera reviewed (approved):
This seems to be working fine and the code looks good to me, @draganescu do you have any concerns for bringing this in?

draganescu commented on packages/block-library/src/media-text/editor.scss:
```diff
@@ -26,11 +26,16 @@
 }
+
+.wp-block-media-text--placeholder-image {
+	min-height: 205px;
```
What is 205?

carolinan replied:
I am trying to match the height of the media area placeholder that shows the three options: upload, media library, featured image.
That way, if no featured image is set, and the featured image option is selected, there will not be a shift in height.
When all three options are available, the height of it is "205.59"

MaggieCabrera replied:
let's add a comment explaining that, for future reference

draganescu commented on packages/block-library/src/media-text/index.php:
```diff
@@ -0,0 +1,65 @@
+		$media_size_slug = $attributes['mediaSizeSlug'];
+	}
+	$processor->set_attribute( 'src', esc_url( $current_featured_image ) );
+	$processor->set_attribute( 'alt', esc_attr( $processor->get_attribute( 'alt' ) ) );
```
Maybe there is a tag processor quirk I don't know, but this line reads like replace alt with the same alt.

carolinan replied:
I don't remember the reasoning behind this. Unless, it needs to be escaped?

carolinan replied:
It is already escaped, I think the line can be removed.

draganescu commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -145,9 +150,42 @@ function MediaTextEdit( { attributes, isSelected, setAttributes } ) {
+		: '';
+	const featuredImageAlt = useFeaturedImage
+		? featuredImageMedia?.alt_text
+		: '';
```
Since we check for `featuredImageURL` and `featuredImageAlt` as `bool` in many places can we set them to `false` if they have no value? IMO the way JS says ` false || ''` is `false`  is better if we default to testing for `false`? Is there any downside?

Alternatively we could make `hasFeaturedImageAlt` and `hasFeaturedImageURL`, respectively. set to ` !== '' `, then use these in `bool` checks later?

carolinan replied:
I tried this, but it meant that the alt text was set to the string "false" instead of being empty when:

- The featured image is not used.
- The current media has no alt text of its own.

It shows the text "false" inside the alt text option in the editor, and in the attribute.

[So this condition on line 316 also needs to be updated.
`value={ mediaAlt || featuredImageAlt }`](https://github.com/WordPress/gutenberg/pull/51491/files#diff-799bc423023028d1e733982e9d106233470b262e34737b34c559ab7b14e001d7R316)

carolinan replied:
@draganescu I think I did not fully understand what to do here.

draganescu replied:
> It shows the text "false" inside the alt text option in the editor, and in the attribute.

It means we need the empty string, In which case it means we should not check for `value={ mediaAlt || featuredImageAlt }` later but instead for `value={ mediaAlt || featuredImageAlt !== '' }`. That's it :)

draganescu reviewed (commented):
I don't want to block this but it would be great if we could address my small code nits:

- a magic CSS number
- avoiding implicit type juggling
- setting an alt value with the same contents as before

Also it should be rebased before testing. @carolinan I could push these too, lmk! 🙇🏻

carolinan committed:
Merge branch 'trunk' into add/media-text-featured-image

carolinan committed:
Remove set_attribute 'alt' from index.php.

carolinan committed:
Merge branch 'trunk' into add/media-text-featured-image

carolinan committed:
Update editor.scss

carolinan committed:
Merge branch 'trunk' into add/media-text-featured-image

t-hamano commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -145,9 +150,42 @@ function MediaTextEdit( { attributes, isSelected, setAttributes } ) {
+	const toggleUseFeaturedImage = () => {
+		setAttributes( {
+			imageFill: false,
+			mediaType: 'image',
```
If we set `mediaType`, an `img` element with no attributes will be saved. This markup is invalid because the `img` element requires at least a `src` attribute.

![media-text](https://github.com/WordPress/gutenberg/assets/54422211/fccbce30-5608-4f1f-b501-b66a5a7bbc4e)

However, I do know that the `img` tag needs to be present in the markup in order to dynamically assign the `src` attribute during dynamic rendering. I am wondering if there is any good approach to solve this problem.

t-hamano replied:
By setting `mediaType` to `undefined`, we will be able to remove the `img` tag that does not have attributes from the saved markup. Instead, I believe that it's possible to inject an `img` tag in the dynamic rendering when `useFeaturedImage` is `true`.

We should be able to use `preg_replace` to inject the `img` element as a string right after the opening tag of the `figure` element. This approach is also used in [the Image block](https://github.com/WordPress/gutenberg/blob/b8474c83aa37c7ab299b4f251a0f96ad598c5f95/packages/block-library/src/image/index.php#L230). In the future, if the HTML API becomes able to inject arbitrary elements, it will be possible to refactor it at that time.

carolinan replied:
Thank you.
I'm trying, but I can't get it to work, the image tag is only rended on the front. I must be missing a step, or there is something that we did not consider.
I will push the changes though so they can be tested.

carolinan replied:
I mean, is is it just because the <figure> is not identical in the editor and front?

t-hamano commented on packages/block-library/src/media-text/media-container.js:
```diff
@@ -110,13 +129,17 @@ function MediaContainer( props, ref ) {
 	const { toggleSelection } = useDispatch( blockEditorStore );
 
-	if ( mediaUrl ) {
+	if ( mediaUrl || featuredImageURL || props.useFeaturedImage ) {
```
```suggestion
	if ( mediaUrl || featuredImageURL || useFeaturedImage ) {
```

t-hamano commented on packages/block-library/src/media-text/index.php:
```diff
@@ -0,0 +1,64 @@
+			$position = round( $attributes['focalPoint']['x'] * 100 ) . '% ' . round( $attributes['focalPoint']['y'] * 100 ) . '%';
+		} else {
+			$position = '50% 50%';
+		}
```
```suggestion
		$position = '50% 50%';
		if ( isset( $attributes['focalPoint'] ) ) {
			$position = round( $attributes['focalPoint']['x'] * 100 ) . '% ' . round( $attributes['focalPoint']['y'] * 100 ) . '%';
		}
```

We can reduce the amount of code by just one line. Additionally, it would match the style of this passage.

https://github.com/WordPress/gutenberg/pull/51491/files#diff-e65105b1f1e325d53864a6ee67246b9f13e65d6fc359ede052dd925c647843adR41-R44

t-hamano reviewed (commented):
I happened to see this PR, but I think this feature makes sense. Maybe I don't understand the context, but I'd like to leave a comment.

carolinan committed:
Update packages/block-library/src/media-text/index.php

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

carolinan committed:
Merge branch 'trunk' into add/media-text-featured-image

carolinan committed:
Update packages/block-library/src/media-text/media-container.js

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

carolinan committed:
Merge branch 'trunk' into add/media-text-featured-image

carolinan committed:
Try removing the image mediaType and inserting the img tag in index.php.

carolinan committed:
CS fix

Adjust spacing.

draganescu reviewed (commented):
@carolinan @t-hamano instead of setting the mediaType to `undefined` which would need the `MediaContainer` to be updated, and also seems wrong because media type for featured _image_ is `image`, we could avoid saving the image in the save function if useFeaturedImage is true:

- so keep `mediaType: 'image'` in edit
- update save to:

```
let image = ! useFeaturedImage ? (
		<img
			src={ mediaUrl }
			alt={ mediaAlt }
			className={ imageClasses || null }
		/>
	) : null;
```

what do you think?

carolinan committed:
Merge branch 'trunk' into add/media-text-featured-image

carolinan commented:
I tried it with the save function update. It is an improvement because the empty img tag is not present if the featured image option is set when there is no featured image on the post.

But in this scenario, the `<img/>` is still in the block markup:

1. The post has a featured image set.
2. A media & text block is added with the option "use featured image".
3. The post is saved.
4. The option "use feature image" is toggled off, but no new media is selected.

carolinan committed:
Add condition to save.js.

t-hamano commented:
> - so keep `mediaType: 'image'` in edit
> - update save to:

This approach seems better. If `mediaType` was set to `undefined` when "Use featured image" was turned on, the image became invisible on the editor and the "Crop image to fill" control etc. also disappeared 😅

> But in this scenario, the `<img/>` is still in the block markup:

This would be an unexpected scenario. We may need to adjust the conditional statement in the save.js function.

colorful-tones commented:
Based on a recent Editor Triage async session - I'm moving this off the [WordPress 6.5 Editor Tasks board](https://github.com/orgs/WordPress/projects/148/views/1).

carolinan committed:
Merge branch 'trunk' into add/media-text-featured-image

carolinan committed:
Merge branch 'trunk' into add/media-text-featured-image

carolinan committed:
Merge branch 'trunk' into add/media-text-featured-image

carolinan committed:
CS: Remove the truthy default value for withIllustration in media-text/media-container.js

carolinan committed:
Update the condition that outputs the img tag

Only output the img tag if there is a mediaURL.
The mediaURL is required because img tags without a src attribute is invalid HTML.

carolinan commented:
I changed the condition in save.js from checking if a featured image is used, to checking if there is a `mediaURL`. This seemed logical because the issue is that an `<img>` tag without a `src` is invalid HTML. And from my manual testing, it seems to solve the issue.

draganescu commented:
I think this is a good state, any improvement can come in from of a follow up. Congratulations everyone for the tweaking and effort to see this through.

Also I forgot to paste the list of contributors
```
Co-authored-by: carolinan <poena@git.wordpress.org>
Co-authored-by: draganescu <andraganescu@git.wordpress.org>
Co-authored-by: MaggieCabrera <onemaggie@git.wordpress.org>
Co-authored-by: t-hamano <wildworks@git.wordpress.org>
Co-authored-by: colorful-tones <colorful-tones@git.wordpress.org>
Co-authored-by: paaljoachim <paaljoachim@git.wordpress.org>
Co-authored-by: annezazu <annezazu@git.wordpress.org>
```

carolinan commented:
Thank you.

ellatrix commented:
These PHP changes need to be backported right? A Trac ticket and core PR would be great!

carolinan commented:
Hmm I think all these PHP changes are part of package updates?

I am looking at the description in https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/blocks/require-dynamic-blocks.php
